### DISCUSSION TITLE
G7 erratic behaviour; defective sensors? bluetooth interference?

### DISCUSSION THREAD

**Discussion by dagord** [NightscoutFoundation/xDrip#3454] [Q&A]
Hello, I recently switched to Dexcom G7 after having happily used G6 with xDrip for a couple of years on the same phone as now (Xiaomi Redmi note 10)
First of all, let me clarify I'm currently running the last nightly version (2024.04.19) and I'm using the
[settings suggested](https://navid200.github.io/xDrip/docs/G6-Recommended-Settings.html])

The first sensor worked fine without noticing any relevant issue until its expiration.

The second sensor gave me a lot of problems!
- after the warmup, which I did on evening, it started returning values every 15 minutes; only on the following morning I got it working as supposed (reading every 5 minutes) without any intervention
- But almost every night, even if at a random time, the 15 minutes issue came up again during the night; once I could fix it forcing a new pair request (forget the bluetooth device -> changing the transmitter id -> insert correct trasmitter id -> pair device again); another night I had to toggle the "avoid scanning" option
- This night I was so frustrated that I decided to remove the sensor and to start a new one.

So now I'm with my 3rd sensor; it worked fine until this afternoon: I went away from my smartphone for 30 minutes and when I come back the 15 minutes thing came up again! 
This is an extract from my logs:
![Screenshot_2024-05-10-17-10-09-656_com eveningoutpost dexdrip](https://github.com/NightscoutFoundation/xDrip/assets/21258057/882987f9-2025-4684-9880-fc59521b6742)

So I decided to test the trasmitter signal with [BLE Scanner](https://navid200.github.io/xDrip/docs/Bluetooth-Scanner.html); both the transmitters (the new and the old, that was near to me and wasn't expired yet) were correctly found but even when "clear and bold" the values were not returned to xDrip; suspecting an interference with the old sensor I moved it to another room and the issue disappeared!

So, coming back to the original issue: why am I experiencing those disconnections during nightime?
I keep my supply of sensors in the bedside table, but I don't think they're causing interference since they're still sealed and not activated.
It was a sensor issue for sure or is there something I can do to investigate further?

I'm asking this because I experienced an identical issue with my old G6 transmitter: after an Android 13 update it started returning values every 15 minutes, but the issue affected only xDrip; the workaround was to use the BYODA Dexcom app and xDrip in companion mode. Eventually, the issue disappeared after an xDrip update.

**Comment by Navid200** [comment]
Xiaomi is a known problem phone.

This is what is recommended for all phones that experience intermittent connectivity:
https://navid200.github.io/xDrip/docs/Intermittent.html

I had occasional disconnects myself with G6 and a Motorola phone until I got a Pixel phone about 9 months ago.
Since then, there has been no disconnects.  I have been using G7 for the past 6 months.

Since you have not posted a screenshot of your status page, I cannot make any other suggestions.

**Reply by dagord** [reply]
Hello. Thanks for your quick reply
I'm attaching the requested screenshot. Please note that at the moment the sensor is working correctly.
![Screenshot_2024-05-10-17-35-41-436_com eveningoutpost dexdrip](https://github.com/NightscoutFoundation/xDrip/assets/21258057/2e7b6918-5bcf-4af9-b8ec-ee4ddf38be88)
![Screenshot_2024-05-10-17-35-46-089_com eveningoutpost dexdrip](https://github.com/NightscoutFoundation/xDrip/assets/21258057/9a7c0de5-3a20-4acf-993a-691fac058976)

I know Xiaomi can be "problematic" and I already read the suggestion for intermittent connectivity.
I was using the same smartphone when I was on Dexcom G6 and I never encountered such issues, at least after the updated mentioned in my post.
I'm wondering if the presence of other unexpired sensors nearby can lead to interferences.

**Reply by Navid200** [reply]
No, the G7 devices that are still inside the applicator and have not been inserted are not activated yet and cannot interfere.  However, the used G7 devices can.

**Reply by dagord** [reply]
Maybe an interference with my old G6 transmitter? It's not expired yet and it lives in the same drawer next to my bed. But it has been unpaired from my phone

**Reply by Navid200** [reply]
My previous statement can be misleading.
Let me clarify.
Having used G7 or any other devices nearby cannot cause an established link to be broken.

I have my used G7 devices in the same room I sleep in and I never experience a disconnect.

When establishing connectivity to a G7 after inserting it, we recommend to do it if possible away from used G7 devices until the connection is established and you have approved a pair request or have had readings.
After that, there should be no problem.

But, if you are connected and it disconnects, either the settings are wrong or the phone interferes with xDrip maintaining connectivity.

**Reply by dagord** [reply]
Hello,
I'm wondering if, under certain circumstances, the bt connection can suffer interferences; e.g. I move to another room, the phone misses some readings, then it founds an unexpired transmitter nearby and it starts with the disconnect/reconnect loop. I'm just guessing.
By the way, I moved my old G6 transmitter and my G7 sensor (which I need to return to Dexcom) to the cellar and I haven't suffered any nightly disconnection anymore. Sometimes the phone still misses a reading, but it works again on the next cycle.
I know it sounds weird; maybe different phone brands act differently in similar circumstance.
Ora maybe it's just a case and I'm really dealing with defective sensors; anyway I preferred to report my experience.

**Reply by Navid200** [reply]
Have you disabled minimize scanning?

I am looking at the possibility of improving the G7 collector.

**Reply by Navid200** [reply]
Can you also make sure that the following setting is enabled?
Settings -> Less common settings -> Aggressive service restarts

**Reply by dagord** [reply]
Both settings are currently enabled.
I'll disable "mimimize scanning" and I'll report if I notice improvements in bt stability

**Reply by Navid200** [reply]
I have opened a PR that I hope will improve things: https://github.com/NightscoutFoundation/xDrip/pull/3461
My hope is that it will not increase battery consumption.  I am testing it now.

**Reply by dagord** [reply]
Hello,
just a quick update.
I've disabled the "Minimize scanning" option 3 days ago. The connection seems quite stable; it still misses some readings (even with the phone at my side) but the connection comes back within 10/15 minutes... but please consider that in the meanwhile I got rid of the unexpired transmitters, so I'm not affected by the potential interferencies anymore.
I've noticed something interesting:
- occasionally, if the reading is missed a the 5th minute, the next reading may occur before the 10th minute (e.g. at the 8th minute; please see the screenshot below); it's not clear to me if it's actually the transmitter sending values to the app or if it's the app requesting values from the transmitter

![exdrip](https://github.com/NightscoutFoundation/xDrip/assets/21258057/52a9fee3-0a9b-409e-8b75-eccf9ea82041)

- this afternoon, xDrip reported that the last reading was 7 minutes prior; I checked the status page which reported "last connected: 2 minutes ago", suggesting that the bluetooth link was up and working but the value was not updated

![Screenshot_2024-05-16-16-05-26-318_com eveningoutpost dexdrip](https://github.com/NightscoutFoundation/xDrip/assets/21258057/ea3e6b44-e8ea-43b8-b21b-4f670a9430fb)
![Screenshot_2024-05-16-16-05-30-718_com eveningoutpost dexdrip](https://github.com/NightscoutFoundation/xDrip/assets/21258057/6178535f-578d-4c88-8497-5377dc7ca16c)

I updated xDrip yesterday so I'm currently running the 2024-05-12 nigthly release.
I think it would be useful to record in the logs the actual timestamps of the readings
Hope this helps!

**Reply by Navid200** [reply]
You can see the timestamps in the data table: https://navid200.github.io/xDrip/docs/Datatables.html

It is great to know that disabling minimize scanning improves performance.  Keep an eye on battery consumption.

xDrip updated the timestamp to when receiving data off grid, which made sense for a G6 because that only happened when switching to a new transmitter.
With G7, G7 starts transmitting once a minute instead of once every 5 minutes after about 15 minutes of disconnect.
xDrip incorrectly updated the timestamp then.  This is incorrect because G7 only deviates when there is a disconnect but the effective grid remains at once every 5 minutes.

This change in behavior from G6 to G7 needs to be accommodated in xDrip, which we are working on.
So, there will be an update to xDrip to address this so that the timestamps are not shifted around every time G7 goes to the once a minute transmission.

**Reply by dagord** [reply]
I haven't experienced any noticeable battery drain on the smartphone so far.
Or are you referring to the sensor life expectation?

**Reply by Navid200** [reply]
No, when you disable minimize scanning, the phone battery consumption may go up.
Since it is not for you, that is perfect.
No, there is no impact on the G7 battery.

**Reply by Navid200** [reply]
Update xDrip to the May 22, 2024 release.
It may help your phone reconnect faster.

**Reply by dagord** [reply]
Hello
I installed the last nightly as soon as it became available.
I'll test it a few days and I'll report my experience!
Thank you, and keep up the good work! :)

**Reply by dagord** [reply]
Hello @Navid200 
I'm sorry but I have to report that with the last 2024-05-22 nigtlhy the connection seems more unstable.
I'm experiencing several "missing signal" alerts (with the smartphone close to me), meaning that the app doesn't get readings for more than 20 minutes. I don't remember such issues with the 2024-05-12 nightly (yes, there were missed readings but for shorter periods).
The behaviour is inconsistend: sometimes the connection is stable even when I'm in a different room, so I don't know if the problem is in the sensor or in the app itself.
By the way, during the prolonged disconnections, the sensor is showing up as stable on BLE scanner (if this makes sense...)

**Reply by Navid200** [reply]
Please update to the Nightly that was released today and tell us if it makes any difference:
https://github.com/NightscoutFoundation/xDrip/releases/tag/2024.05.28

**Reply by dagord** [reply]
Hello again
I installed the last nightly and the BT connection is clearly more stable! Thank you!
Anyway, the's still a bug that I had already noticed previously: when a reading is missed (e.g. I'm in another room) the subsequent reading is often skipped as well (even if I'm back), so I need to wait for other 5 minutes. It's not a big issue but it could be useful to investigate it.

**Reply by Navid200** [reply]
I cannot recreate this problem with a Pixel 6a:
![Screenshot_20240603-112121](https://github.com/NightscoutFoundation/xDrip/assets/51497406/fb91940e-522c-4c30-9cf7-dd2ab1248977)  ![Screenshot_20240603-112228](https://github.com/NightscoutFoundation/xDrip/assets/51497406/fdd6c092-19b8-467a-99a1-1a3b70fdcb23)  
  
You can see only one backfill here:  
![Screenshot_20240603-112257](https://github.com/NightscoutFoundation/xDrip/assets/51497406/b5bc0932-26aa-43cb-850c-85b8110b1c7b)
  
  
G7 transmits only for 15 seconds.  You are not going to get a reading if you are back not exactly when G7 is transmitting.

**Reply by Navid200** [reply]
Does the problem still persist?

**Reply by Navid200** [reply]
> Anyway, the's still a bug that I had already noticed previously: when a reading is missed (e.g. I'm in another room) the subsequent reading is often skipped as well (even if I'm back), so I need to wait for other 5 minutes. 

Would you please help me understand this? 

Do you mean after you return to the phone, you will never have proper operation and you will have a missed reading once every 10 minutes?  Meaning you only get every other reading consistently?

Or, when you return, you will have one more missed reading before everything goes back to normal and you have no more missed readings until you move away from the phone again?

**Reply by dagord** [reply]
The second version is correct. For example, let's suppose I go to the other room at the 4th minute, so the phone skips the reading at the 5th minute; at the 6th minute I come back, but the phone also skips the next cycle. At the 15th minute, it becomes regular again. It's not a regular behaviour, so it's difficult to replicate it.

I discovered, however, that by re-enabling the 'minimize scanning' option (which you had advised me to disable: https://github.com/NightscoutFoundation/xDrip/discussions/3454#discussioncomment-9422116), the connection has become much more stable, at least with the latest nightly release (2024.06.28); I don't know if this makes sense

**Reply by Navid200** [reply]
OK,
Thanks for explaining.  And I am sorry about asking you to disable minimize scanning.  It was only an experiment.  I am glad you re-enabled it.

I have occasionally experienced the same thing.  What I consider a serious problem is if there are consistent disconnects, not when I walk away and on return, miss one more reading.  

Just to let you know, there is an open PR: https://github.com/NightscoutFoundation/xDrip/pull/3567  
The description of it suggests that it is to address the problem you have described.  

Thanks

**Reply by dagord** [reply]
> And I am sorry about asking you to disable minimize scanning. It was only an experiment. I am glad you re-enabled it.

No problem! I'm happy if this can help you in improving the app. It seems that disabling minimize scanning actually improved the connection on the old release, but on the newer releases it just breaks something.
And yes: my issue is exactly the one described in the thread you linked!
